From: Attila Szalay <sasa@ubainba.hu>
Date: Sat, 3 Sep 2022 22:45:40 +0100
Subject: Replace direct md5sum file access with dpkg --verify calls

Closes: #905680
---
 contrib/syslog-ng-debun | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/contrib/syslog-ng-debun b/contrib/syslog-ng-debun
index 7a7bb2e..49c5f0b 100755
--- a/contrib/syslog-ng-debun
+++ b/contrib/syslog-ng-debun
@@ -963,7 +963,10 @@ debun_extra_debian () {
 	printf "\nDebian specific checks\n"
 	printf "Check package files integrity\n"
 	cd /
-	md5sum --quiet -c /var/lib/dpkg/info/syslog-ng*.md5sums && printf "Package files are intact\n"
+	for package in $(dpkg -l syslog-ng\* | grep "ii" | awk -F " " '{print $2}')
+	do
+		dpkg --verify ${package} && printf "Package ${package} files are intact\n"
+	done
 	printf "list syslog-related packages\n"
 	dpkg -l |grep syslog > ${tmpdir}/deb.packages
 }
