From: Kokan <kokaipeter@gmail.com>
Date: Fri, 24 Jul 2020 10:34:39 +0200
Subject: syslogformat: reorder functions

Signed-off-by: Kokan <kokaipeter@gmail.com>
Origin: https://github.com/syslog-ng/syslog-ng/commit/2fb6a11160ec87f4d49bce16097a2c559fceb527
---
 modules/syslogformat/syslog-format.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/modules/syslogformat/syslog-format.c b/modules/syslogformat/syslog-format.c
index 6d53a32..8c70159 100644
--- a/modules/syslogformat/syslog-format.c
+++ b/modules/syslogformat/syslog-format.c
@@ -47,6 +47,14 @@ static struct
   NVHandle raw_message;
 } handles;
 
+static inline void
+sd_step(const guchar **data, gint *left)
+{
+  (*data)++;
+  (*left)--;
+}
+
+
 static gboolean
 log_msg_parse_pri(LogMessage *self, const guchar **data, gint *length, guint flags, guint16 default_pri)
 {
@@ -499,14 +507,6 @@ log_msg_parse_hostname(LogMessage *self, const guchar **data, gint *length,
   *length = left;
 }
 
-
-static inline void
-sd_step(const guchar **data, gint *left)
-{
-  (*data)++;
-  (*left)--;
-}
-
 /**
  * log_msg_parse:
  * @self: LogMessage instance to store parsed information into
