From: SZALAY Attila <sasa@debian.org>
Date: Fri, 2 Sep 2016 19:06:13 +0200
Subject: Fixed the name of the libmongoc pc file

Signed-off-by: SZALAY Attila <sasa@debian.org>
---
 configure.ac | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure.ac b/configure.ac
index 611e7b5..4d517cc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -999,9 +999,9 @@ AC_DEFINE_UNQUOTED(ENABLE_LEGACY_MONGODB_OPTIONS, [$enable_legacy_mongodb_option
 
 if test "x$with_mongoc" = "xauto"; then
   with_mongoc="system"
-  PKG_CHECK_MODULES(LIBMONGO, libmongoc >= $LMC_MIN_VERSION,,with_mongoc="auto-internal")
+  PKG_CHECK_MODULES(LIBMONGO, libmongoc-1.0 >= $LMC_MIN_VERSION,,with_mongoc="auto-internal")
 elif test "x$with_mongoc" = "xsystem"; then
-  PKG_CHECK_MODULES(LIBMONGO, libmongoc >= $LMC_MIN_VERSION)
+  PKG_CHECK_MODULES(LIBMONGO, libmongoc-1.0 >= $LMC_MIN_VERSION)
 fi
 
 if test "x$with_mongoc" = "xinternal" || test "x$with_mongoc" = "xauto-internal"; then
