From 997d1eb069301d0c0dc8ce78a98133c0acb11153 Mon Sep 17 00:00:00 2001
From: Antal Nemes <antal.nemes@balabit.com>
Date: Wed, 23 Jan 2019 16:19:35 +0100
Subject: [PATCH] cmake: add missing detection for O_LARGEFILE

Signed-off-by: Antal Nemes <antal.nemes@balabit.com>
---
 CMakeLists.txt        | 8 ++++++++
 syslog-ng-config.h.in | 1 +
 2 files changed, 9 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1e099f3991..9a0f3cd101 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -73,6 +73,7 @@ include(GenerateYFromYm)
 include(CheckStructHasMember)
 
 add_definitions(-D_GNU_SOURCE=1)
+add_definitions(-D_LARGEFILE64_SOURCE=1)
 
 include(CheckSockaddrStorage)
 if (HAVE_STRUCT_SOCKADDR_STORAGE)
@@ -92,6 +93,13 @@ if (ENABLE_IPV6)
     set(SYSLOG_NG_ENABLE_IPV6 ${HAVE_IPV6})
 endif()
 
+set(CMAKE_REQUIRED_DEFINITIONS "-D_LARGEFILE64_SOURCE=1")
+set(CMAKE_EXTRA_INCLUDE_FILES "fcntl.h")
+check_type_size(O_LARGEFILE O_LARGEFILE)
+if (HAVE_O_LARGEFILE)
+  set(SYSLOG_NG_HAVE_O_LARGEFILE 1)
+endif()
+
 check_symbol_exists(strtoll stdlib.h SYSLOG_NG_HAVE_STRTOLL)
 check_symbol_exists(strnlen string.h SYSLOG_NG_HAVE_STRNLEN)
 check_symbol_exists(strtok_r string.h SYSLOG_NG_HAVE_STRTOK_R)
diff --git a/syslog-ng-config.h.in b/syslog-ng-config.h.in
index 465b7a3622..81e046c25f 100644
--- a/syslog-ng-config.h.in
+++ b/syslog-ng-config.h.in
@@ -60,3 +60,4 @@
 #cmakedefine01 SYSLOG_NG_HAVE_INOTIFY
 #cmakedefine01 SYSLOG_NG_HAVE_GETRANDOM
 #cmakedefine01 SYSLOG_NG_USE_CONST_IVYKIS_MOCK
+#cmakedefine01 SYSLOG_NG_HAVE_O_LARGEFILE
