diff --git a/lib/tlscontext.c b/lib/tlscontext.c
index dc34f2c..8eb09ce 100644
--- a/lib/tlscontext.c
+++ b/lib/tlscontext.c
@@ -298,6 +298,9 @@ tls_context_setup_session(TLSContext *self)
       if (self->key_file && self->cert_file && !SSL_CTX_check_private_key(self->ssl_ctx))
         goto error;
 
+      if (file_exists(self->ca_file) && !SSL_CTX_load_verify_locations(self->ssl_ctx, self->ca_file, NULL))
+        goto error;
+
       if (file_exists(self->ca_dir) && !SSL_CTX_load_verify_locations(self->ssl_ctx, NULL, self->ca_dir))
         goto error;
 
@@ -383,6 +386,7 @@ tls_context_free(TLSContext *self)
   g_free(self->key_file);
   g_free(self->cert_file);
   g_free(self->ca_dir);
+  g_free(self->ca_file);
   g_free(self->crl_dir);
   g_free(self->cipher_suite);
   g_free(self);
diff --git a/lib/tlscontext.h b/lib/tlscontext.h
index 98a2f85..7ceec51 100644
--- a/lib/tlscontext.h
+++ b/lib/tlscontext.h
@@ -68,6 +68,7 @@ struct _TLSContext
   gchar *key_file;
   gchar *cert_file;
   gchar *ca_dir;
+  gchar *ca_file;
   gchar *crl_dir;
   gchar *cipher_suite;
   SSL_CTX *ssl_ctx;
diff --git a/modules/afsocket/afsocket-grammar.ym b/modules/afsocket/afsocket-grammar.ym
index 9a4ccf0..0402f19 100644
--- a/modules/afsocket/afsocket-grammar.ym
+++ b/modules/afsocket/afsocket-grammar.ym
@@ -155,6 +155,7 @@ afinet_grammar_set_dest_driver(AFInetDestDriver *dd)
 %token KW_KEY_FILE
 %token KW_CERT_FILE
 %token KW_CA_DIR
+%token KW_CA_FILE
 %token KW_CRL_DIR
 %token KW_TRUSTED_KEYS
 %token KW_TRUSTED_DN
@@ -663,6 +664,11 @@ tls_option
 	    last_tls_context->ca_dir = g_strdup($3);
             free($3);
           }
+	| KW_CA_FILE '(' string ')'
+	  {
+	    last_tls_context->ca_file = g_strdup($3);
+	    free($3);
+	  }
 	| KW_CRL_DIR '(' string ')'
 	  {
 	    last_tls_context->crl_dir = g_strdup($3);
diff --git a/modules/afsocket/afsocket-parser.c b/modules/afsocket/afsocket-parser.c
index f9e14a4..f749f7e 100644
--- a/modules/afsocket/afsocket-parser.c
+++ b/modules/afsocket/afsocket-parser.c
@@ -48,6 +48,7 @@ static CfgLexerKeyword afsocket_keywords[] = {
   { "key_file",           KW_KEY_FILE },
   { "cert_file",          KW_CERT_FILE },
   { "ca_dir",             KW_CA_DIR },
+  { "ca_file",            KW_CA_FILE },
   { "crl_dir",            KW_CRL_DIR },
   { "trusted_keys",       KW_TRUSTED_KEYS },
   { "trusted_dn",         KW_TRUSTED_DN },
