--- a/lib/tlscontext.c
+++ b/lib/tlscontext.c
@@ -48,6 +48,7 @@
   gchar *dhparam_file;
   gchar *pkcs12_file;
   gchar *ca_dir;
+  gchar *ca_file;
   gchar *crl_dir;
   gchar *cipher_suite;
   gchar *ecdh_curve_list;
@@ -691,6 +692,9 @@
         goto error;
     }
 
+  if (_is_file_accessible(self, self->ca_file) && !SSL_CTX_load_verify_locations(self->ssl_ctx, NULL, self->ca_file))
+    goto error;
+
   if (_is_file_accessible(self, self->ca_dir) && !SSL_CTX_load_verify_locations(self->ssl_ctx, NULL, self->ca_dir))
     goto error;
 
@@ -788,6 +792,7 @@
   g_free(self->cert_file);
   g_free(self->dhparam_file);
   g_free(self->ca_dir);
+  g_free(self->ca_file);
   g_free(self->crl_dir);
   g_free(self->cipher_suite);
   g_free(self->ecdh_curve_list);
@@ -972,6 +977,13 @@
 }
 
 void
+tls_context_set_ca_file(TLSContext *self, const gchar *ca_file)
+{
+  g_free(self->ca_file);
+  self->ca_file = g_strdup(ca_file);
+}
+
+void
 tls_context_set_crl_dir(TLSContext *self, const gchar *crl_dir)
 {
   g_free(self->crl_dir);
--- a/lib/tlscontext.h
+++ b/lib/tlscontext.h
@@ -119,6 +119,7 @@
 void tls_context_set_cert_file(TLSContext *self, const gchar *cert_file);
 void tls_context_set_pkcs12_file(TLSContext *self, const gchar *pkcs12_file);
 void tls_context_set_ca_dir(TLSContext *self, const gchar *ca_dir);
+void tls_context_set_ca_file(TLSContext *self, const gchar *ca_file);
 void tls_context_set_crl_dir(TLSContext *self, const gchar *crl_dir);
 void tls_context_set_cipher_suite(TLSContext *self, const gchar *cipher_suite);
 void tls_context_set_ecdh_curve_list(TLSContext *self, const gchar *ecdh_curve_list);
--- a/modules/afsocket/afsocket-grammar.ym
+++ b/modules/afsocket/afsocket-grammar.ym
@@ -176,6 +176,7 @@
 %token KW_DHPARAM_FILE
 %token KW_PKCS12_FILE
 %token KW_CA_DIR
+%token KW_CA_FILE
 %token KW_CRL_DIR
 %token KW_TRUSTED_KEYS
 %token KW_TRUSTED_DN
@@ -790,6 +791,11 @@
 	    tls_context_set_ca_dir(last_tls_context, $3);
             free($3);
           }
+	| KW_CA_FILE '(' string ')'
+	  {
+	    tls_context_set_ca_file(last_tls_context, $3);
+	    free($3);
+	  }
 	| KW_CRL_DIR '(' string ')'
 	  {
 	    tls_context_set_crl_dir(last_tls_context, $3);
--- a/modules/afsocket/afsocket-parser.c
+++ b/modules/afsocket/afsocket-parser.c
@@ -24,7 +24,7 @@
 #include "afsocket.h"
 #include "driver.h"
 #include "cfg-parser.h"
-#include "afsocket-grammar.h"
+#include <afsocket-grammar.h>
 
 extern int afsocket_debug;
 
@@ -50,6 +50,7 @@
   { "dhparam_file",       KW_DHPARAM_FILE },
   { "pkcs12_file",        KW_PKCS12_FILE },
   { "ca_dir",             KW_CA_DIR },
+  { "ca_file",            KW_CA_FILE },
   { "crl_dir",            KW_CRL_DIR },
   { "trusted_keys",       KW_TRUSTED_KEYS },
   { "trusted_dn",         KW_TRUSTED_DN },
--- a/modules/afsocket/Makefile.am
+++ b/modules/afsocket/Makefile.am
@@ -42,8 +42,8 @@
 module_LTLIBRARIES				+= modules/afsocket/libafsocket.la
 modules_afsocket_libafsocket_la_CPPFLAGS	=	\
 	$(AM_CPPFLAGS) $(libsystemd_CFLAGS)	\
-	-I${top_srcdir}/modules/afsocket		\
-	-I${top_builddir}/modules/afsocket
+	-I${top_builddir}/modules/afsocket	\
+	-I${top_srcdir}/modules/afsocket
 
 modules_afsocket_libafsocket_la_LIBADD	=	\
 	$(MODULE_DEPS_LIBS) $(OPENSSL_LIBS)	\
