diff --git a/modules/afsocket/afinet.c b/modules/afsocket/afinet.c
index f83ff38..61893a4 100644
--- a/modules/afsocket/afinet.c
+++ b/modules/afsocket/afinet.c
@@ -496,8 +496,7 @@
     }
 
 
-  if ((self->bind_ip && !resolve_hostname(&self->super.bind_addr, self->bind_ip)) ||
-      !resolve_hostname(&self->super.dest_addr, self->super.hostname))
+  if ((self->bind_ip && !resolve_hostname(&self->super.bind_addr, self->bind_ip)))
     return FALSE;
 
   afinet_set_port(self->super.dest_addr, self->dest_port ? : default_dest_port, self->super.flags & AFSOCKET_DGRAM ? "udp" : "tcp");
@@ -531,6 +530,9 @@
 {
   AFInetDestDriver *self = (AFInetDestDriver *) s;
 
+  if (!resolve_hostname(&self->super.dest_addr, self->super.hostname))
+    return FALSE;
+
   return afinet_setup_socket(fd, self->super.dest_addr, (InetSocketOptions *) s->sock_options_ptr, AFSOCKET_DIR_SEND);
 }
 
@@ -551,7 +553,7 @@
 
           saved_caps = g_process_cap_save();
           g_process_cap_modify(CAP_NET_RAW, TRUE);
-          self->lnet_ctx = libnet_init(self->super.dest_addr->sa.sa_family == AF_INET ? LIBNET_RAW4 : LIBNET_RAW6, NULL, error);
+          self->lnet_ctx = libnet_init(self->super.bind_addr->sa.sa_family == AF_INET ? LIBNET_RAW4 : LIBNET_RAW6, NULL, error);
           g_process_cap_restore(saved_caps);
           if (!self->lnet_ctx)
             {
@@ -686,7 +688,10 @@
 #if ENABLE_SPOOF_SOURCE
   AFInetDestDriver *self = (AFInetDestDriver *) s;
 
-  if (self->spoof_source && self->lnet_ctx && msg->saddr && (msg->saddr->sa.sa_family == AF_INET || msg->saddr->sa.sa_family == AF_INET6))
+  /* NOTE: this code should probably become a LogTransport instance so that
+   * spoofed packets are also going through the LogWriter queue */
+
+  if (self->spoof_source && self->lnet_ctx && msg->saddr && (msg->saddr->sa.sa_family == AF_INET || msg->saddr->sa.sa_family == AF_INET6) && log_writer_opened((LogWriter *) self->super.writer))
     {
       gboolean success = FALSE;
 
diff --git a/modules/afsocket/afsocket.c b/modules/afsocket/afsocket.c
index ae9c5c2..5b192f4 100644
--- a/modules/afsocket/afsocket.c
+++ b/modules/afsocket/afsocket.c
@@ -893,7 +893,7 @@
 
   if ((self->flags & AFSOCKET_SYSLOG_PROTOCOL) == 0)
     {
-      switch (self->dest_addr->sa.sa_family)
+      switch (self->bind_addr->sa.sa_family)
         {
         case AF_UNIX:
           source = !!(self->flags & AFSOCKET_STREAM) ? SCS_UNIX_STREAM : SCS_UNIX_DGRAM;
@@ -1109,6 +1109,8 @@
       return FALSE;
     }
 
+  g_assert(self->dest_addr);
+
   rc = g_connect(sock, self->dest_addr);
   if (rc == G_IO_STATUS_NORMAL)
     {
@@ -1165,7 +1167,6 @@
   /* these fields must be set up by apply_transport, so let's check if it indeed did */
   g_assert(self->transport);
   g_assert(self->bind_addr);
-  g_assert(self->dest_addr);
   g_assert(self->hostname);
   g_assert(self->dest_name);
 
